src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -0,0 +1,251 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+using SafeCurlHandle = Interop.libcurl.SafeCurlHandle;
+using CURLoption = Interop.libcurl.CURLoption;
+using CURLcode = Interop.libcurl.CURLcode;
+using CURLINFO = Interop.libcurl.CURLINFO;
+
+namespace System.Net.Http
+{
+    internal class CurlHandler : HttpMessageHandler
+    {
+        #region Constants
+        private const string UriSchemeHttps = "https";
+        #endregion
+
+        #region Fields
+        private volatile bool _anyOperationStarted;
+        private volatile bool _disposed;
+        private bool _automaticRedirection = true;
+        #endregion
+
+        internal CurlHandler()
+        {
+        }
+
+        #region Properties
+        internal bool AutomaticRedirection
+        {
+            get
+            {
+                return _automaticRedirection;
+            }
+
+            set
+            {
+                CheckDisposedOrStarted();
+                _automaticRedirection = value;
+            }
+        }
+        #endregion
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && !_disposed)
+            {
+                _disposed = true;
+            }
+
+            base.Dispose(disposing);
+        }
+
+        protected internal override Task<HttpResponseMessage> SendAsync(
+            HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException("request", SR.net_http_handler_norequest);
+            }
+            if (request.RequestUri.Scheme == UriSchemeHttps)
+            {
+                throw NotImplemented.ByDesignWithMessage("HTTPS stack is not yet implemented");
+            }
+            if (request.Content != null)
+            {
+                throw NotImplemented.ByDesignWithMessage("HTTP requests with a body are not yet supported");
+            }
+
+            CheckDisposed();
+
+            SetOperationStarted();
+
+            // Create RequestState object and save current values of handler settings.
+            RequestState state = new RequestState();
+            state.CancellationToken = cancellationToken;
+            state.RequestMessage = request;
+
+            try
+            {
+                Task.Factory.StartNew(
+                    StartRequest,
```

Comment:
Is this same as the second part of the comment about inheriting from TaskCompletionSource? Could you explain a bit about how to avoid the delegate creation?